src/dodal/devices/panda_fast_grid_scan.py

```diff
@@ -0,0 +1,145 @@
+    y_steps: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y_NUM_STEPS")
+    z_steps: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z_NUM_STEPS")
+
+    x_step_size: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "X_STEP_SIZE")
+    y_step_size: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y_STEP_SIZE")
+    z_step_size: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z_STEP_SIZE")
+
+    # This value is fixed by the time between X steps detector deadtime. The only reason it is a PV
+    # Is so the value can be read by the motion program in the PPMAC
+    time_between_x_steps_ms = Component(EpicsSignalWithRBV, "TIME_BETWEEN_X_STEPS")
+
+    run_up_distance: EpicsSignalWithRBV = Component(EpicsSignal, "RUNUP_DISTANCE")
+
+    x_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "X_START")
+    y1_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y_START")
+    y2_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y2_START")
+    z1_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z_START")
+    z2_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z2_START")
+
+    position_counter: EpicsSignalRO = Component(EpicsSignalRO, "Y_COUNTER")
+    scan_invalid: EpicsSignalRO = Component(EpicsSignalRO, "SCAN_INVALID")
+
+    run_cmd: EpicsSignal = Component(EpicsSignal, "RUN.PROC")
+    stop_cmd: EpicsSignal = Component(EpicsSignal, "STOP.PROC")
+    status: EpicsSignalRO = Component(EpicsSignalRO, "SCAN_STATUS")
+
+    expected_images: Signal = Component(Signal)
+
+    # Kickoff timeout in seconds
+    KICKOFF_TIMEOUT: float = 5.0
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        def set_expected_images(*_, **__):
+            x, y, z = self.x_steps.get(), self.y_steps.get(), self.z_steps.get()
+            first_grid = x * y
+            second_grid = x * z
+            self.expected_images.put(first_grid + second_grid)
+
+        self.x_steps.subscribe(set_expected_images)
+        self.y_steps.subscribe(set_expected_images)
+        self.z_steps.subscribe(set_expected_images)
+
+    def is_invalid(self) -> bool:
+        if "GONP" in self.scan_invalid.pvname:
+            return False
+        return self.scan_invalid.get()
+
+    def kickoff(self) -> StatusBase:
+        # Check running already here?
+        st = DeviceStatus(device=self, timeout=self.KICKOFF_TIMEOUT)
+
+        def scan():
+            try:
+                self.log.debug("Running scan")
+                self.run_cmd.put(1)
+                self.log.debug("Waiting for scan to start")
+                await_value(self.status, 1).wait()
+                st.set_finished()
+            except Exception as e:
+                st.set_exception(e)
+
+        threading.Thread(target=scan, daemon=True).start()
+        return st
+
+    def complete(self) -> DeviceStatus:
+        return GridScanCompleteStatus(self)
```

Comment:
Must: This status looks at `position_counter` and waits for it to be the same as `set_expected_images` so I don't think it will ever complete for the panda.

Reply:
Thanks, I'd removed this after https://github.com/DiamondLightSource/dodal/pull/279#discussion_r1452384685, but the status did something slightly different